### PR TITLE
Make GovernorCountingFractional accept partial votes

### PR DESCRIPTION
**Pull request by davidlaprade**
Implements #13

### CODE REVIEW

File: src/GovernorCountingFractional.sol [hunk 1]
```diff
@@ -134,23 +146,28 @@ abstract contract GovernorCountingFractional is Governor {
...
         uint256 proposalId,
-        uint256 weight,
+        address account,
+        uint128 weight,
         bytes memory voteData
     ) internal {
         require(voteData.length == 48, "GovernorCountingFractional: invalid voteData");
 
-        (uint128 forVotes, uint128 againstVotes, uint128 abstainVotes) = _decodePackedVotes(voteData);
+        (uint128 _forVotes, uint128 _againstVotes, uint128 _abstainVotes) = _decodePackedVotes(voteData);
```
Comment (mds1) [inline src/GovernorCountingFractional.sol:155]: _(arbitrary comment location to start a thread)_

One small comment is that this PR would have been easier to review without the variable name changes (i.e. adding a leading underscore), since lots of lines had changes like this so I had to see if that was the only change on the line or not. Ideally we should handle things like that in their own PR to keep diffs simpler, not a big deal though of course 🙂

File: src/GovernorCountingFractional.sol [hunk 2]
```diff
@@ -134,23 +146,28 @@ abstract contract GovernorCountingFractional is Governor {
...
+        uint128 _existingWeight = _proposalVotersWeightCast[proposalId][account];
 
         require(
-            uint256(forVotes) + againstVotes + abstainVotes <= uint128(weight),
-            "GovernorCountingFractional: votes exceed weight"
+            uint256(_forVotes) + _againstVotes + _abstainVotes + _existingWeight <= weight,
+            "GovernorCountingFractional: vote would exceed weight"
         );
 
+        _proposalVotersWeightCast[proposalId][account] = _existingWeight + _forVotes + _againstVotes + _abstainVotes;
```
Comment (mds1) [inline src/GovernorCountingFractional.sol:164]: Could save that sum to a var instead of computing it twice

**Review by mds1 — COMMENTED** [review]
looks great!

File: test/GovernorCountingFractional.t.sol [hunk 3]
```diff
@@ -165,8 +171,18 @@ contract GovernorCountingFractionalTest is Test {
...
+  {
+    vm.assume(_addr > address(0));
+    assumeNoPrecompiles(_addr);
+    vm.label(_addr, _name);
+    return _addr;
+  }
+
   function _randomAddress(uint256 salt1) public view returns (address) {
-    return address(uint160(uint256(keccak256(abi.encodePacked(salt1, block.number)))));
+    return address(uint160(uint256(keccak256(abi.encodePacked(salt1, block.timestamp)))));
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:185]: Oh, interesting, I didn't realize that `block.timestamp` was locked in tests. Thanks, I like that approach

File: test/GovernorCountingFractional.t.sol [hunk 4]
```diff
@@ -249,35 +265,37 @@ contract GovernorCountingFractionalTest is Test {
...
 
-  function _castVotes(Voter[4] memory voters, uint256 _proposalId) internal {
-    Voter memory voter;
-    for (uint8 _i = 0; _i < voters.length; _i++) {
-      voter = voters[_i];
-
-      assert(!governor.hasVoted(_proposalId, voter.addr));
+  function _castVotes(Voter memory _voter, uint256 _proposalId) internal {
+    if (_voter.weight == 0) return;
+    assert(!governor.hasVoted(_proposalId, _voter.addr));
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:270]: To be honest, I don't remember exactly what I wanted when I added these -- it was a while ago. I'll revisit

File: test/GovernorCountingFractional.t.sol [hunk 5]
```diff
@@ -454,33 +499,42 @@ contract GovernorCountingFractionalTest is Test {
     voter.addr = _randomAddress(_weight);
     // The weight cannot overflow the max supply for the token, but must overflow the
     // max for the GovernorFractional contract.
-    voter.weight = bound(_weight, MAX_VOTE_WEIGHT, token.exposed_maxSupply());
+    voter.weight = bound(_weight, MAX_VOTE_WEIGHT + 1, token.exposed_maxSupply());
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:502]: Per the comment above, we need a weight that is greater than the max voting weight for the governor

File: test/GovernorCountingFractional.t.sol [hunk 6]
```diff
@@ -454,33 +499,42 @@ contract GovernorCountingFractionalTest is Test {
...
   function testFuzz_ParamLengthIsChecked(
     uint256 _weight,
     FractionalVoteSplit memory _voteSplit,
-    uint256 invalidParamLength
+    uint256 _invalidParamLength
   ) public {
-    invalidParamLength = bound(invalidParamLength, 1, 256);
-    vm.assume(invalidParamLength != 48);
-    bytes memory fractionalVoteData = new bytes(invalidParamLength);
+    _invalidParamLength = bound(_invalidParamLength, 1, 255);
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:526]: Actually, this number should be much higher. Users could send any crazy large amount of data. I'll update

File: test/GovernorCountingFractional.t.sol [hunk 7]
```diff
@@ -454,33 +499,42 @@ contract GovernorCountingFractionalTest is Test {
...
+    vm.assume(_invalidParamLength != 48);
+
+    // Construct a byte array of arbitrary length and arbitrary values.
+    bytes memory _invalidVoteData = new bytes(_invalidParamLength);
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:530]: Wow yeah, for some reason I didn't think you could fuzz over bytes. This is way better, thanks!

File: test/GovernorCountingFractional.t.sol [hunk 8]
```diff
@@ -599,4 +654,273 @@ contract GovernorCountingFractionalTest is Test {
...
+    address _voterAddr,
+    uint256 _weight,
+    uint256 _votePercentage1,
+    uint256 _votePercentage2,
+    uint256 _votePercentage3,
+    FractionalVoteSplit memory _voteSplit
+  ) public {
+    // Build the vote split.
+    _voteSplit = _randomVoteSplit(_voteSplit);
+    _votePercentage1 = bound(_votePercentage1, 0.0e18, 1.0e18);
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:668]: The setup is confusing, I agree. I had to read a bunch of stuff to refresh myself upon stepping back in here too. I'll add some more comments

File: test/GovernorCountingFractional.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (mds1) [reply test/GovernorCountingFractional.t.sol:502]: Oh whoops, I was thinking the range here had a min of `_weight` and max of `MAX_VOTE_WEIGHT + 1` 🙃 Yes makes sense

File: test/GovernorCountingFractional.t.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (mds1) [reply test/GovernorCountingFractional.t.sol:526]: Going to mark this resolved in favor of just using a `bytes` fuzzer input as discussed in https://github.com/ScopeLift/flexible-voting/pull/31#discussion_r1088109807

File: test/GovernorCountingFractional.t.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (mds1) [reply test/GovernorCountingFractional.t.sol:185]: You of course can change the timestamp with `vm.warp`, but yea by default no blocks are mined and no time elapses unless you use a cheatcode to do so

File: test/GovernorCountingFractional.t.sol [hunk 9]
```diff
@@ -599,4 +654,273 @@ contract GovernorCountingFractionalTest is Test {
...
+    assertEq(_firstVote.forVotes + _secondVote.forVotes + _thirdVote.forVotes, _actualForVotes);
+    assertEq(
+      _firstVote.againstVotes + _secondVote.againstVotes + _thirdVote.againstVotes,
+      _actualAgainstVotes
+    );
+    assertEq(
+      _firstVote.abstainVotes + _secondVote.abstainVotes + _thirdVote.abstainVotes,
+      _actualAbstainVotes
+    );
+  }
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:763]: I feel like this should be enough. Given [5000 fuzz runs](https://github.com/ScopeLift/flexible-voting/blob/debfa653e2e18d2421c9aa405db228baee58f8a1/foundry.toml#L11) the odds seem pretty good that we end up with at least one in which all of the weight is cast.

To make sure, though, I'll add a couple non-fuzz tests that specify the percentages so that we certainly will run tests that exercise both full weight and partial weight

File: src/GovernorCountingFractional.sol [hunk 10]
```diff
@@ -98,13 +98,17 @@ abstract contract GovernorCountingFractional is Governor {
         uint256 weight,
         bytes memory voteData
     ) internal virtual override {
-        require(!_proposalVotersHasVoted[proposalId][account], "GovernorCountingFractional: vote already cast");
-        _proposalVotersHasVoted[proposalId][account] = true;
+        require(weight > 0, "GovernorCountingFractional: no weight");
+        if (_proposalVotersWeightCast[proposalId][account] == weight) {
+          revert("GovernorCountingFractional: vote already cast");
+        }
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:104]: > Why not make this also a require for consistency?

It was a require and I removed it in 455cc1f4a2430f96decec6e4952faebbbc2f4185 because I found it confusing. I think this is more self-documenting.

> should we tweak the error string now that you can vote multiple times?

Sure, that's fair

File: src/GovernorCountingFractional.sol (hunk 1 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:155]: Noted. Though I'll be honest that I made all of these changes because I expected to get comments on it if I didn't update the old code to fit the current style guide

File: src/GovernorCountingFractional.sol (hunk 2 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:164]: Thanks, nice catch!

File: test/GovernorCountingFractional.t.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Done in 575f596

File: test/GovernorCountingFractional.t.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Done in 575f596

File: test/GovernorCountingFractional.t.sol [hunk 11]
```diff
@@ -249,35 +265,37 @@ contract GovernorCountingFractionalTest is Test {
...
+      vm.expectEmit(true, false, false, true);
+      emit VoteCast(_voter.addr, _proposalId, _voter.support, _voter.weight, "Yay");
+    }
+
+    vm.prank(_voter.addr);
+    governor.castVoteWithReasonAndParams(_proposalId, _voter.support, "Yay", fractionalizedVotes);
 
-      vm.prank(voter.addr);
-      governor.castVoteWithReasonAndParams(_proposalId, voter.support, "Yay", fractionalizedVotes);
+    assert(governor.hasVoted(_proposalId, _voter.addr));
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Done in 575f596

File: test/GovernorCountingFractional.t.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Removed in 575f596

File: test/GovernorCountingFractional.t.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Done in 575f596

File: test/GovernorCountingFractional.t.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Done in 575f596

File: test/GovernorCountingFractional.t.sol [hunk 12]
```diff
@@ -599,4 +654,273 @@ contract GovernorCountingFractionalTest is Test {
     assertEq(_againstVotes, _actualAgainstVotes);
     assertEq(_abstainVotes, _actualAbstainVotes);
   }
+
+  function testFuzz_CanCastPartialWeightMultipleTimes(
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:669]: Done in 575f596

File: test/GovernorCountingFractional.t.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol:778]: Done in 575f596

File: test/GovernorCountingFractional.t.sol [hunk 13]
```diff
@@ -599,4 +654,273 @@ contract GovernorCountingFractionalTest is Test {
...
+  function testFuzz_NominalVotingCannotExceedOverallWeightWithMultipleVotes(
+    uint256[4] memory _weights
+  ) public {
+    Voter memory _voter = _setupNominalVoters(_weights)[0];
+    _mintAndDelegateToVoter(_voter);
+
+    uint256 _proposalId = _createAndSubmitProposal();
+    bytes memory _emptyDataBecauseWereVotingNominally;
+
+    vm.expectEmit(true, false, false, true);
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Done in 575f596

File: test/GovernorCountingFractional.t.sol [hunk 14]
```diff
@@ -599,4 +654,273 @@ contract GovernorCountingFractionalTest is Test {
...
+      uint128(_voter.weight.mulWadDown(_voteSplit.percentAbstain)) // abstainVotes
+    );
+
+    // Mint, delegate, and propose.
+    _mintAndDelegateToVoter(_voter);
+    uint256 _proposalId = _createAndSubmitProposal();
+
+    if (_isCastingNominallyFirst) {
+      // Vote nominally. It should succeed.
+      vm.expectEmit(true, false, false, true);
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Done in 575f596

File: src/GovernorCountingFractional.sol [hunk 15]
```diff
@@ -41,7 +41,7 @@ abstract contract GovernorCountingFractional is Governor {
      * @dev See {IGovernor-hasVoted}.
      */
     function hasVoted(uint256 proposalId, address account) public view virtual override returns (bool) {
-        return _proposalVotersHasVoted[proposalId][account];
+        return _proposalVotersWeightCast[proposalId][account] > 0;
```
[the remark this answers is not in the document]
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:44]: Done in 575f596

File: src/GovernorCountingFractional.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol]: Error string tweaked in 575f596

File: src/GovernorCountingFractional.sol (hunk 2 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol]: Done in 575f596

File: test/GovernorCountingFractional.t.sol [hunk 16]
```diff
@@ -249,35 +264,37 @@ contract GovernorCountingFractionalTest is Test {
...
 
-  function _castVotes(Voter[4] memory voters, uint256 _proposalId) internal {
-    Voter memory voter;
-    for (uint8 _i = 0; _i < voters.length; _i++) {
-      voter = voters[_i];
-
-      assert(!governor.hasVoted(_proposalId, voter.addr));
+  function _castVotes(Voter memory _voter, uint256 _proposalId) internal {
+    if (_voter.weight == 0) return;
+    assertEq(governor.hasVoted(_proposalId, _voter.addr), false);
```
Comment (mds1) [inline test/GovernorCountingFractional.t.sol:269]: Nit, you can use `assertFalse(governor.hasVoted(_proposalId, _voter.addr))` instead. Feel free to resolve this comment yourself.

File: test/GovernorCountingFractional.t.sol [hunk 17]
```diff
@@ -249,35 +264,37 @@ contract GovernorCountingFractionalTest is Test {
...
+      vm.expectEmit(true, true, true, true);
+      emit VoteCast(_voter.addr, _proposalId, _voter.support, _voter.weight, "Yay");
+    }
+
+    vm.prank(_voter.addr);
+    governor.castVoteWithReasonAndParams(_proposalId, _voter.support, "Yay", fractionalizedVotes);
 
-      vm.prank(voter.addr);
-      governor.castVoteWithReasonAndParams(_proposalId, voter.support, "Yay", fractionalizedVotes);
+    assertEq(governor.hasVoted(_proposalId, _voter.addr), true);
```
Comment (mds1) [inline test/GovernorCountingFractional.t.sol:292]: Nit, you can use `assertTrue(governor.hasVoted(_proposalId, _voter.addr))` instead. Feel free to resolve this comment yourself.

File: src/GovernorCountingFractional.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (mds1) [reply src/GovernorCountingFractional.sol]: Fair enough, lgtm!

File: src/GovernorCountingFractional.sol (hunk 1 above)
Comment (mds1) [reply src/GovernorCountingFractional.sol:155]: Personally I don't like the leading underscore convention, so you wouldn't have gotten comments from me 😛 (I find it adds a lot of noise and hurts readability without adding much value, and there's other ways to solve the the same problem anyway)

Comment (mds1) [reply src/GovernorCountingFractional.sol:155]: Either way, resolving 👌

**Review by mds1 — APPROVED** [review]

File: src/GovernorCountingFractional.sol (hunk 15 above)
Comment (apbendi) [inline src/GovernorCountingFractional.sol:44]: Hmm, I think there's a subtle issue here that I'm a little worried about: currently if `hasVoted` returns `true` there is an implicit assumption that they won't vote again and their votes won't change. We're changing that, and I could see it breaking tooling or extension contracts that make that assumption. Not sure what if any action we can take is, though. Hmmm.

File: src/GovernorCountingFractional.sol [hunk 18]
```diff
@@ -98,13 +98,17 @@ abstract contract GovernorCountingFractional is Governor {
         uint256 weight,
         bytes memory voteData
     ) internal virtual override {
-        require(!_proposalVotersHasVoted[proposalId][account], "GovernorCountingFractional: vote already cast");
-        _proposalVotersHasVoted[proposalId][account] = true;
+        require(weight > 0, "GovernorCountingFractional: no weight");
+        if (_proposalVotersWeightCast[proposalId][account] == weight) {
```
Comment (apbendi) [inline src/GovernorCountingFractional.sol:102]: Might be better to defensively check `>=`

File: src/GovernorCountingFractional.sol [hunk 19]
```diff
@@ -134,23 +146,28 @@ abstract contract GovernorCountingFractional is Governor {
...
 
-        (uint128 forVotes, uint128 againstVotes, uint128 abstainVotes) = _decodePackedVotes(voteData);
+        (uint128 _forVotes, uint128 _againstVotes, uint128 _abstainVotes) = _decodePackedVotes(voteData);
 
-        require(
-            uint256(forVotes) + againstVotes + abstainVotes <= uint128(weight),
-            "GovernorCountingFractional: votes exceed weight"
-        );
+        uint128 _existingWeight = _proposalVotersWeightCast[proposalId][account];
+        uint256 _newWeight = uint256(_forVotes) + _againstVotes + _abstainVotes + _existingWeight;
```
Comment (apbendi) [inline src/GovernorCountingFractional.sol:158]: So, if I understand how this implementation works, the call to castVote submits the marginal votes (i.e. the new votes) to the contract. This works, but I'm trying to think through the implications for the delegate contract, which now needs to track how much weight has or hasn't been submitted, or else look it up from the Governor.

The alternative approach would be to assume the delegate contract always votes with 100% of votes each time, and then have our Governor check that the For/Against/Abstain values for the current vote are less than the For/Against/Abstain values for the previous vote.

Not sure one is better than the other just trying to make sure we've considered the tradeoffs. What are your thoughts?

**Review by apbendi — COMMENTED** [review]
Looking good @davidlaprade, thanks!

File: src/GovernorCountingFractional.sol (hunk 15 above)
Comment (mds1) [reply src/GovernorCountingFractional.sol:44]: Interesting point. The [docs](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/bc6de21fe229520b33d13a4d60bbeb1bc5f91bcd/contracts/governance/IGovernor.sol#L185-L189) for `hasVoted` are (understandly) ambiguous on this point:

> Returns whether `account` has cast a vote on `proposalId`.

So we aren't breaking that definition. But to your point people almost certainly have interpreted this as a binary "all or none", since there was no flex voting until now. I think the options are:
1. Do nothing, since this is technically fine. Confirm this is ok and won't break much by searching for some on/off-chain `hasVoted` usage via sourcegraph
2. Add a new `hasFinishedVoting` that returns true if you've voted with all your weight, to ensure there's a method with the original implied semantics
3. Restore the implied semantics of this method and add a `hasPartiallyVoted` method 

I'm personally ok with any of these, perhaps we pull this out into a follow up issue and ask OZ what they'd prefer, to increases chances of getting this upstreamed?

File: src/GovernorCountingFractional.sol (hunk 19 above)
Comment (mds1) [reply src/GovernorCountingFractional.sol:158]: Good callout, I don't think we really discussed and I think this approach was just implied.

Seems the current approach is cheaper because we don't currently track an individual user's split, whereas voting with the new full total amount would require us to track that for each user, so I think you'd end up doing more reading/writing than currently

File: src/GovernorCountingFractional.sol (hunk 15 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:44]: Yeah, it's a good point. I'm also not sure what to do about this. Having this return `false` until all weight is cast seems confusing as well -- maybe even more so.

File: src/GovernorCountingFractional.sol (hunk 19 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:158]: > This works, but I'm trying to think through the implications for the delegate contract, which now needs to track how much weight has or hasn't been submitted, or else look it up from the Governor

Does it? I don't think that's necessarily true. For example, when we remove `CAST_VOTE_WINDOW` in our flex voting AToken in #18 I don't think we'll need to do it. As long as the client/delegate contract knows how many votes any given user is owed for a given proposal, and it does this internal accounting properly, that should be sufficient.

Just to make this concrete, here's an example:
* Alice deposits 25 UNI into a flex voting aToken
* Bob deposits 60 UNI
* Charlie deposits 15 UNI
* so 100 UNI has been deposited in total now
* Dave borrows 80 UNI
* the aToken's UNI balance is now 20 UNI
* a UNI proposal is issued
* Alice expresses her FOR vote to the aToken
* Bob expresses an AGAINST vote to the aToken
* someone calls `aToken.castVote`, rolling these two votes up together and submitting them fractionally to the governor
  * to do this, the aToken asks the governance token what its balance (i.e. voting weight) was at the proposal snapshot, it is 20 UNI
  * the aToken knows that Alice is responsible for 25 of the 100 total UNI deposited, so it casts 5 FOR votes for her (25% of its available 20 UNI voting weight)
  * the aToken knows that Bob is responsible for 60 of the 100 total UNI deposited, so it casts 12 AGAINST votes for him (60% of its available 20 UNI voting weight)
* sometime later, Charlie expresses a FOR vote to the aToken, then calls `aToken.castVote`
  * the aToken doesn't need to lookup its remaining votes, since it knows that Charlie is responsible for 15 of the 100 total UNI deposited, so it casts 3 FOR votes for him (15% of its available 20 UNI voting weight)

File: src/GovernorCountingFractional.sol (hunk 15 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:44]: Curious what @mds1 thinks of this issue

File: test/GovernorCountingFractional.t.sol (hunk 16 above)
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Thanks. Done in b202e74

File: test/GovernorCountingFractional.t.sol (hunk 17 above)
Comment (davidlaprade) [reply test/GovernorCountingFractional.t.sol]: Thanks. Done in b202e74

File: src/GovernorCountingFractional.sol (hunk 18 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol]: Thanks. Done in b202e74

File: src/GovernorCountingFractional.sol (hunk 19 above)
Comment (mds1) [reply src/GovernorCountingFractional.sol:158]: > This works, but I'm trying to think through the implications for the delegate contract, which now needs to track how much weight has or hasn't been submitted, or else look it up from the Governor

Ah I overlooked this and didn't address it. I agree with @davidlaprade here though, I don't think this has downsides for the delegate contract (aToken, in this example). Instead of (from the above example) this flow:

- Alice expresses her FOR vote to the aToken
- Bob expresses an AGAINST vote to the aToken
- someone calls aToken.castVote, rolling these two votes up together and submitting them fractionally to the governor

You can also have the delegate contract designed as follows:

- Alice casts her FOR vote to the aToken, by calling `aToken.castVote()` which forwards her vote to the governor
- Bob expresses an AGAINST vote to the aToken, by calling `aToken.castVote()` which forwards his vote to the governor

In either flow, the aToken contract just needs to have one storage write per vote (whether my cast vote or david's express vote) indicating alice voted. Currently, it does a storage write anyway to save off the expressed vote counts, so you just change what gets stored

Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:158]: Yup agreed, that would also work

Comment (apbendi) [reply src/GovernorCountingFractional.sol:158]: @davidlaprade in your example above, when Charlie calls `atToken.castVote`, how does the aToken know to submit only his vote, and not to resbumit Alice and Bob's?

Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:167]: > how does the aToken know to submit only his vote, and not to resbumit Alice and Bob's

It would have to clear Alice and Bob's votes from storage after submitting them

Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:167]: Okay, I implemented both approaches so that we can compare:
* [incremental partial voting](https://github.com/ScopeLift/flexible-voting/pull/33) (i.e. what this PR assumed)
* [replacement parial voting](https://github.com/ScopeLift/flexible-voting/pull/34) (i.e. the alternative Ben pointed out)

Grain of salt: the Governor diffs aren't a fair gauge of complexity since they are both based on this PR and this PR already implemented incremental partial voting in the Governor, so that diff for #33 is going to be smaller. The better place to compare complexity is in the client, i.e. the changes to `ATokenFlexVoting`. And there I think you'll see that the two are basically identical in complexity.

With respect to gas costs, the two functions we care about are`expressVote` and `castVote`. Here's how they compare between the two implementations:

| function | implementation | min | avg | median | max |
|----------|----------------|-----|-----|--------|-----|
| expressVote | replacement | 2467            | 36498  | 47776  | 48320  |
| expressVote | incremental | 2489            | 37316  | 47798  | 48342  |
| castVote       | replacement | 12076           | 66037  | 70596  | 108396 |
| castVote       | incremental | 12053           | **49295**  | **53887**  | **68207**  |

That last row is pretty dramatic.

One last data point, I asked @jferas @wildmolasses and @garyghayrat the following question:

> If I told you that a governance system allowed you to cast partial votes, how would you assume that system tallied the votes you cast? More concretely: if you had 20 weight to cast, and you voted 3 times -- first by casting 3 votes, next by casting 7 votes, and third by casting 10 votes -- how much of your voting weight would you assume had been used up in the process?

Everyone assumed that in a scenario like the one described, you would have cast all of your weight with these votes, i.e. the default assumption was that the governor would treat partial votes incrementally.

By no means is this conclusive, but I think it does say something about ease of integration that 5/6 assumed the incremental behavior was how this would work.

Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:167]: As a note: @apbendi, @mds1, and I just got on a call and discussed this and agreed that the incremental approach is probably best. It's cheaper and doesn't seem like it has a downside for the client.

An idempotent approach (like the replacement approach Ben suggested) would work really well in an environment where you have to worry about lock contention. But that isn't a concern here, since we effectively hold a global lock within the context of each transaction.

File: src/GovernorCountingFractional.sol (hunk 15 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:53]: @rsolari (since I know you discussed this feature with @apbendi) I'm curious if you have any take on this question.

TLDR: how should `hasVoted` work if it's possible for people to cast partial votes?

Some options:
1. it could return true so long as some weight has been cast, even if not all weight has been cast
1. it could return false unless all weight has been cast
1. some combination of (1) and then also add another function like `hasFullyVoted` to capture the behavior that `hasVoted` used to have
1. something else… we’re all ears

What would be best from your (Tally's) point of view

Comment (rsolari) [reply src/GovernorCountingFractional.sol:53]: Tally doesn't use `hasVoted()` right now, so I don't have a strong opinion from Tally's perspective.

I can imagine a frontend using `hasVoted()` to see whether a user is still able to vote. I could also imagine a subgraph/indexer to using it to count up votes cast. 

For a Flexible Voter, both an indexer and a frontend would need to know how many votes are cast so far. We can't answer that with a boolean, so I don't think option 3 really helps. I'd suggest option 1 because it keeps the semantics from `IGovernor`: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/91e8d0ba3c3beb8a1db31310d8599664e48639ef/contracts/governance/IGovernor.sol#L187

I'd suggest adding a helper e.g. `votesCast(uint256 proposalId, address account)` for getting the votes so far. Tally would just use an indexer to avoid RPC calls, but it's nice to give the option to smart contracts and frontends.

Comment (mds1) [reply src/GovernorCountingFractional.sol:53]: @davidlaprade mind pulling this out into a separate issue, then we can mark resolved?

File: src/GovernorCountingFractional.sol (hunk 19 above)
Comment (mds1) [reply src/GovernorCountingFractional.sol:167]: Probably can resolve this thread now?

**Review by mds1 — APPROVED** [review]

File: src/GovernorCountingFractional.sol (hunk 15 above)
Comment (davidlaprade) [reply src/GovernorCountingFractional.sol:53]: Awesome, thanks so much for the input @rsolari!

I think I agree that adding a new helper would be the best approach here. Tracking this in https://github.com/ScopeLift/flexible-voting/issues/37

**Review by apbendi — APPROVED** [review]
